sympy/solvers/ode.py

```diff
@@ -1482,6 +1491,226 @@ def _test_term(coeff, order):
+        e = (e.subs(x, match['mobius'])).subs(x, x**match['k'])
+        if B:
+            e1 = Integral(A/2, x)
+            e1 = logcombine(e, force=True)
+            e = simplify(e - e1)
+        sol = exp(e)*(sol*x**((-match['k']+1)/2))
+        sol = Eq(func, sol)
+
+    # if sol is None then we can try for series solution
+    if sol is None:
+        try:
+            sol = dsolve(eq, func, hint = "2nd_power_series_regular", x0=x0)
+        except ValueError:
+            pass
+    if sol is None:
+        raise NotImplementedError("....")
+
+    return sol
+
+def match_2nd_2F1_hypergeometric(I, k, sing_point, func, r):
+
+    from sympy import factor
+    x = func.args[0]
+    a = Wild("a")
+    b = Wild("b")
+    c = Wild("c")
+    alpha = Wild("alpha")
+    beta = Wild("beta")
+    gamma = Wild("gamma")
+    delta = Wild("delta")
+
+    rn = {'type':None}
+    # I0 of the standerd 2F1 equation.
+    I0 = ((a-b+1)*(a-b-1)*x**2 + 2*((1-a-b)*c + 2*a*b)*x + c*(c-2))/(4*x**2*(x-1)**2)
+    if sing_point != [0, 1]:
+        # If singular point is [0, 1] then we have standerd equation.
+        eqs = []
+        sing_eqs = [-delta/gamma, -beta/alpha, (delta-beta)/(alpha-gamma)]
+        # making equations for the finding the mobius transformation
+        for i in range(3):
+            if i<len(sing_point):
+                eqs.append(Eq(sing_eqs[i], sing_point[i]))
+            else:
+                eqs.append(Eq(1/sing_eqs[i], 0))
+        # solving above equations for the mobius transformation
+        _delta = solve(eqs[0], delta)[0]
+        _beta = solve(eqs[1], beta)[0]
+        _gamma = solve(((eqs[2]).subs(beta, _beta)).subs(delta, _delta), gamma)[0]
+        subs = (alpha*x + beta)/(gamma*x + delta)
+        subs = subs.subs(beta, _beta)
+        subs = subs.subs(delta, _delta)
+        subs = subs.subs(gamma, _gamma)
+        subs = cancel(subs)
+    else:
+        subs = x
+    # applying mobius transformation in I0.
+    I0 = I0.subs(x, subs)
+    I0 = I0*(subs.diff(x))**2
+    I0 = factor(I0)
+    dict_I = {x**2:0, x:0, 1:0}
+    I0_num, I0_dem = I0.as_numer_denom()
+    # collecting coeff of (x**2, x), of the standerd equation.
+    dict_I0 = collect(expand(I0_num), [x**2, x], evaluate=False)
+    # collecting coeff of (x**2, x) from I0 of the given equation.
+    dict_I.update(collect(expand(cancel(I*I0_dem)), [x**2, x], evaluate=False))
+    eqs = []
+    # We are comparing the coeff of powers of different x, for finding the values of
+    # parameters of standerd equation.
+    for key, val in dict_I.items():
+        eqs.append(Eq(val, dict_I0[key]))
+    # I am trying to solve the equations for a, b, c.
+    # this is not a efficient way to solve and it is slowing the process
+    # also it is not giving solution always.
+
+    _c = simplify(solve(eqs[2], c)[0])
+    _a, _b = solve([eqs[0].subs(c, _c), eqs[1].subs(c, _c)], [a, b])[0]
+    _c = _c.subs(a, _a)
+    _c = _c.subs(b, _b)
+
```

Comment:
Here, I am solving three equations for a, b, c. But I need a more efficient way to solve.
Because it is slowing and not giving always a solution.

Reply:
```
In [12]: eq=[Eq(2*a**2 + b**2 - 2*b - 4, x**2 + 6*x*y - 4*x*z + y**2 - 4*y*z + 4*z**2 - 4*z - 1), Eq(-2*a**2 + 2*b**2, 2*x
    ...: **2 + 4*x*y - 4*x*z + 2*y**2 - 4*y*z + 4*z - 2), Eq(b**2 + 2*b, a*x**2 - 2*x*y + y**2 - 1)]

In [13]: eq
Out[13]: 
⎡   2    2              2                    2              2                 2      2      2                      2    
⎣2⋅a  + b  - 2⋅b - 4 = x  + 6⋅x⋅y - 4⋅x⋅z + y  - 4⋅y⋅z + 4⋅z  - 4⋅z - 1, - 2⋅a  + 2⋅b  = 2⋅x  + 4⋅x⋅y - 4⋅x⋅z + 2⋅y  - 4

                 2            2            2    ⎤
⋅y⋅z + 4⋅z - 2, b  + 2⋅b = a⋅x  - 2⋅x⋅y + y  - 1⎦

```

Let's take an example, I am getting above equation and I need to find a, b, c in term of x, y, z.

Reply:
I thought part of the point of the method was that it wasn't necessary to solve a system of algebraic equations

Reply:
We are equating the standard Invariant and shifted Invariant(of the given equation) for a, b,c so for this I am getting these equations.

 **4th step of algo from paper**-



"When the equivalence is possible, from the singularities of J0(x) and by c**omparing it with the shifted invariant** of the corresponding pFq equation, compute the parameters {a, b, c} entering the pFq equation
 such that the equivalence exists as well as the parameters {α, β, γ, δ} entering the M ̈obius trans-
formation ."

In [maple](https://www.maplesoft.com/support/help/maple/view.aspx?path=dsolve%2Fhyper3), they are doing it very efficiently and also for very complex equations.

Reply:
So we are comparing equation (30) from the ODE we want to solve with equation (31) which comes from the general form of the 2F1 equation?

If I understand correctly then equation (31) is always the same regardless of which ODE we want to solve. That should mean that we can find the solution for a, b, c in general terms now without needing to call solve each time I think. We first need to rearrange the coefficient in (30) so that it matches the form:
```
In [12]: (alpha*x**2 + beta*x + gamma) / (4*x**2*(x-1)**2)                                               
Out[12]: 
   2          
α⋅x  + β⋅x + γ
──────────────
   2        2 
4⋅x ⋅(x - 1)  
```
To do that I suggest multiplying by `4*x**2*(x-1)**2` and using apart/expand etc.

Then I think we need to solve:
```
In [13]: eqs = [Eq(alpha, (a-b+1)*(a-b-1)), Eq(beta, 2*((1-a-b)*c+2*a*b)), Eq(gamma, c*(c-2))]           

In [14]: eqs                                                                                             
Out[14]: [α = (a - b - 1)⋅(a - b + 1), β = 4⋅a⋅b + 2⋅c⋅(-a - b + 1), γ = c⋅(c - 2)]

In [15]: solve(eqs, [a, b, c]) 
```
I get an answer from solve for that. It finds 8 different solutions which makes sense since the equations are quadratic in 3 variables.

With this hard-coded answer from solve we get the solutions just by substituting for alpha, beta and gamma and maybe simplifying. It might make sense to work through this system of equations more "manually" to find any degenerate equations where the general solution above fails:

We can see that we can always solve the 3rd equation for c and get two distinct values. Also the 1st equation is can always be solved for a-b to get two different values. Putting those together in the 2nd equation we will get a quadratic in either a or b (suppose a). So I guess the only possible degenerate case would be if the coefficient of the `a**2` term is then zero (so it's not really quadratic). I think the coefficient is always 4 though so that isn't possible. So I think that we always get 8 (possibly not distinct) solutions and can somehow choose whichever we think is best without worrying about degenerate cases.

It might be better to explicitly code the steps in the last paragraph as a solver algorithm that only needs to find roots of quadratics and do substitutions.

Have I understood this properly?

Reply:
Where does I come from here?

Reply:
Can you explain how you end up with these equations (from the beginning)?

Reply:
```julia
In [28]: eq=x*(x-1)*f(x).diff(x, 2) + ((a+b+1)*x-c)*f(x).diff(x) + a*b*f(x)

In [23]: A=((a + b+1)*x -c)/(x**2-x)

In [24]: B= b*a/(x**2-x)

In [25]: I = A.diff(x)/2 + A**2/4 - B

In [26]: factor(I)
Out[26]: 
 2  2          2                        2  2              2                  2
a ⋅x  - 2⋅a⋅b⋅x  + 4⋅a⋅b⋅x - 2⋅a⋅c⋅x + b ⋅x  - 2⋅b⋅c⋅x + c  + 2⋅c⋅x - 2⋅c - x 
──────────────────────────────────────────────────────────────────────────────
                                   2        2                                 
                                4⋅x ⋅(x - 1)   
```

the first equation is simply the standard invariant of 2F1 type of equation.

Reply:
That is the coefficient from (31) in a less simple form:
```julia
In [121]: I2 = ((a-b+1)*(a-b-1)*x**2 +2*((1-a-b)*c + 2*a*b)*x + c*(c-2)) / (4*x**2*(x-1)**2)  # (31)     

In [122]: factor(I) == factor(I2)                                                                        
Out[122]: True
```

Reply:
Let 's take the example of the paper eq(28). 
Let's assume by somehow(I will explain later) we transform eq(28) into eq(30). Now our problem is to find the solution of eq(30) then we can easily find the solution of eq(28) by applying the transformation eq(29) to the solution of eq(30). 

Here, I have to transform the 2F1 equation into eq(30) by some transformation.
then again we will apply the same transformation to the solution of 2F1 and we will get the solution of eq(30) and from the solution of eq(30) we will get solution of eq(28)

Now, in eq(30) regular singular point is (0, 1, -1) so from this, we will find Mobius transformation as we have to shift the singular point of standard equation (2F1) from (0, 1 oo) to (0, 1, -1) for equivalence. which is `2*x/(x-1)` for this equation.

Now the table 1 in the paper shows when the Invariant of eq(30)(for general) will match with it we can always transform the 2F1 equation into eq(30) by Mobius. 
 
Now, we will find the transformation by which 2F1 will become of type(30). First, we will have to change the singularity position of 2F1 to the singularity of eq(30) so will normalize the equation and will apply Mobius and it will become in the form of eq(30) and we will compare the transform equation and eq(30) and we will find values of a, b, c. 

But we will not apply these transformations directly to the equation.  We have another way to do it. 
We will find the Invariant of 2F1(coeff of y in 31 after normalizing) and then for Mobius transformation, we can compute Invariant directly by eq(12). So we will compute the Invariant.

Reply:
```

In [29]: eq=x*(x-1)*f(x).diff(x, 2) + ((a+b+1)*x-c)*f(x).diff(x) + a*b*f(x)

In [30]: A=((a + b+1)*x -c)/(x**2-x)

In [31]: B= b*a/(x**2-x)

In [32]: I = A.diff(x)/2 + A**2/4 - B

In [33]: factor(I)
Out[33]: 
 2  2          2                        2  2              2                  2
a ⋅x  - 2⋅a⋅b⋅x  + 4⋅a⋅b⋅x - 2⋅a⋅c⋅x + b ⋅x  - 2⋅b⋅c⋅x + c  + 2⋅c⋅x - 2⋅c - x 
──────────────────────────────────────────────────────────────────────────────
                                   2        2                                 
                                4⋅x ⋅(x - 1)                                  

```
Now, we are using eq(12) to find the Invariant of the transform equation directly.
In [34]: I = ((2*x/(x-1)).diff(x))*I.subs(x, 2*x/(x-1))

```
In [35]: factor(I)
Out[35]: 
 ⎛   2  2                    2                2  2          2              2  2      2      2        2            2⎞ 
-⎝4⋅a ⋅x  - 8⋅a⋅b⋅x - 4⋅a⋅c⋅x  + 4⋅a⋅c⋅x + 4⋅b ⋅x  - 4⋅b⋅c⋅x  + 4⋅b⋅c⋅x + c ⋅x  - 2⋅c ⋅x + c  + 2⋅c⋅x  - 2⋅c - 4⋅x ⎠ 
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
                                                       2        2                                                    
                                                    8⋅x ⋅(x + 1)                                                     



```

And then we will compare it with eq(30) for a, b, c.

Reply:
> That is the coefficient from (31) in a less simple form:
> 
> ```julia
> In [121]: I2 = ((a-b+1)*(a-b-1)*x**2 +2*((1-a-b)*c + 2*a*b)*x + c*(c-2)) / (4*x**2*(x-1)**2)  # (31)     
> 
> In [122]: factor(I) == factor(I2)                                                                        
> Out[122]: True
> ```

coeff of 31 is just the Invariant of Normal equation.

Reply:
In general, let's take the Mobius transformation -
```
x⋅α + β
───────
x⋅γ + δ
```
then we have to compare the eq(30) with -
```
In [48]: I = (((alpha*x + beta)/(gamma*x + delta)).diff(x))*I.subs(x, (alpha*x + beta)/(gamma*x + delta))

In [49]: factor(I)
Out[49]: 
            ⎛ 2  2  2      2          2  2          2  2          2                                                     
(α⋅δ - β⋅γ)⋅⎝a ⋅x ⋅α  + 2⋅a ⋅x⋅α⋅β + a ⋅β  - 2⋅a⋅b⋅x ⋅α  + 4⋅a⋅b⋅x ⋅α⋅γ - 4⋅a⋅b⋅x⋅α⋅β + 4⋅a⋅b⋅x⋅α⋅δ + 4⋅a⋅b⋅x⋅β⋅γ - 2⋅a⋅
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                                        
                                                                                                                        

   2                      2                                                2  2  2      2          2  2          2      
b⋅β  + 4⋅a⋅b⋅β⋅δ - 2⋅a⋅c⋅x ⋅α⋅γ - 2⋅a⋅c⋅x⋅α⋅δ - 2⋅a⋅c⋅x⋅β⋅γ - 2⋅a⋅c⋅β⋅δ + b ⋅x ⋅α  + 2⋅b ⋅x⋅α⋅β + b ⋅β  - 2⋅b⋅c⋅x ⋅α⋅γ -
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                 2                    2                 
                                                                      4⋅(x⋅α + β) ⋅(x⋅α - x⋅γ + β - δ)                  

                                          2  2  2      2          2  2        2            2  2                         
 2⋅b⋅c⋅x⋅α⋅δ - 2⋅b⋅c⋅x⋅β⋅γ - 2⋅b⋅c⋅β⋅δ + c ⋅x ⋅γ  + 2⋅c ⋅x⋅δ⋅γ + c ⋅δ  + 2⋅c⋅x ⋅α⋅γ - 2⋅c⋅x ⋅γ  + 2⋅c⋅x⋅α⋅δ + 2⋅c⋅x⋅β⋅γ 
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                                        
                                                                                                                        

                             2    2  2              2⎞
- 4⋅c⋅x⋅δ⋅γ + 2⋅c⋅β⋅δ - 2⋅c⋅δ  - x ⋅α  - 2⋅x⋅α⋅β - β ⎠
──────────────────────────────────────────────────────


```

Reply:
I think it would be a lot easier if you put these calculations into a notebook.

Since the Mobius transformation is invertible can we not apply it in either direction? We can find the transform that takes (28) to (30) and then use the inverse of that to transform the solution of (30) back:
https://en.wikipedia.org/wiki/M%C3%B6bius_transformation#Formula_for_the_inverse_transformation

Reply:
> So we are comparing equation (30) from the ODE we want to solve with equation (31) which comes from the general form of the 2F1 equation?
> 
> If I understand correctly then equation (31) is always the same regardless of which ODE we want to solve. That should mean that we can find the solution for a, b, c in general terms now without needing to call solve each time I think. We first need to rearrange the coefficient in (30) so that it matches the form:
> 
> ```
> In [12]: (alpha*x**2 + beta*x + gamma) / (4*x**2*(x-1)**2)                                               
> Out[12]: 
>    2          
> α⋅x  + β⋅x + γ
> ──────────────
>    2        2 
> 4⋅x ⋅(x - 1)  
> ```
> 
> To do that I suggest multiplying by `4*x**2*(x-1)**2` and using apart/expand etc.
> 
> Then I think we need to solve:
> 
> ```
> In [13]: eqs = [Eq(alpha, (a-b+1)*(a-b-1)), Eq(beta, 2*((1-a-b)*c+2*a*b)), Eq(gamma, c*(c-2))]           
> 
> In [14]: eqs                                                                                             
> Out[14]: [α = (a - b - 1)⋅(a - b + 1), β = 4⋅a⋅b + 2⋅c⋅(-a - b + 1), γ = c⋅(c - 2)]
> 
> In [15]: solve(eqs, [a, b, c]) 
> ```
> 
> I get an answer from solve for that. It finds 8 different solutions which makes sense since the equations are quadratic in 3 variables.
> 
> With this hard-coded answer from solve we get the solutions just by substituting for alpha, beta and gamma and maybe simplifying. It might make sense to work through this system of equations more "manually" to find any degenerate equations where the general solution above fails:
> 
> We can see that we can always solve the 3rd equation for c and get two distinct values. Also the 1st equation is can always be solved for a-b to get two different values. Putting those together in the 2nd equation we will get a quadratic in either a or b (suppose a). So I guess the only possible degenerate case would be if the coefficient of the `a**2` term is then zero (so it's not really quadratic). I think the coefficient is always 4 though so that isn't possible. So I think that we always get 8 (possibly not distinct) solutions and can somehow choose whichever we think is best without worrying about degenerate cases.
> 
> It might be better to explicitly code the steps in the last paragraph as a solver algorithm that only needs to find roots of quadratics and do substitutions.
> 
> Have I understood this properly?

Actually, we can do this way also. By applying the Mobius transformation into eq(30) to change it into the form of eq(31) then we will compare it.

Reply:
> I think it would be a lot easier if you put these calculations into a notebook.
> 
> Since the Mobius transformation is invertible can we not apply it in either direction? We can find the transform that takes (28) to (30) and then use the inverse of that to transform the solution of (30) back:
> https://en.wikipedia.org/wiki/M%C3%B6bius_transformation#Formula_for_the_inverse_transformation

Yes, we can do the inverse.